### PR TITLE
Enable ADL for math functions

### LINKED ISSUE #855: Math functions are not ADL-friendly
The implementation uses hardcoded functions from `std` in situations like [this](https://github.com/ComputationalRadiationPhysics/alpaka/blob/f3d363bc581e378a4ef6440ec423c7575066b556/include/alpaka/math/exp/ExpStdLib.hpp#L49). It is done consistently for all math functions, I will use `exp` as an example.

This way is not friendly for ADL in case the argument is a user-defined type with `std::is_arithmetic` trait enabled and a custom `exp` implemented, since that implementation would not be found. A [standard approach](https://en.cppreference.com/w/cpp/language/adl) to look for the user-defined first and have `std::exp` as a fallback is to put `using std::exp;` inside the function and then call just `exp`. It looks to me we can do that and I made a toy example which works this way. In terms of implementation, it is awkward for us since we also need the right version of `exp` for return type deduction (before C++14), and it is not allowed to put `using std::exp;` just inside the `struct` (only inside a namespace or function body). So it looks like we would need another wrapper function to resolve all that. And then also to avoid conflicts between our function `exp` and the one we are ADL-looking for.

### PR DESCRIPTION
**Pull request by bernhardmgruber**
Pre-refactoring: Replace static functions of math trait implementations with `operator()`. We need this so we make unqualified calls without name lookup immediately finding the enclosing member function.

All math traits now have a default implementation which does an unqualified call to the corresponding math function. This allows to find overloads for user provided types which are not covered by alpaka or any backend. Integral and floating-point types are covered by the backends.

### CODE REVIEW

File: include/alpaka/math/abs/Traits.hpp
Comment (sbastrakov) [inline include/alpaka/math/abs/Traits.hpp:57]: I think i don't fully understand it. Now a user calls that `alpaka::abs(acc, arg)` declared on line 54 in their kernel. It in turn does ADL on line 40 to switch to user's `abs()` definition when applicable. But when and how do we switch to backend-specific calls such as `struct Abs<AbsUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point<TArg>::value>>` ?

Comment (sbastrakov) [reply include/alpaka/math/abs/Traits.hpp:57]: I think the schema of how these things are organized is worth documenting. As seen from my question, it's probably not that simple to grasp, even though I discovered lack of ADL there and tried a simple fix (with a different way than yours)

File: include/alpaka/math/abs/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/abs/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/acos/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/acos/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/asin/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/asin/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/atan/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/atan/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/atan2/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/atan2/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/cbrt/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/cbrt/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/ceil/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/ceil/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/cos/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/cos/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/erf/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/erf/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/exp/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/exp/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/floor/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/floor/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/fmod/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/fmod/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/log/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/log/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/max/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/max/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/min/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/min/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/trunc/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/trunc/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/tan/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/tan/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/pow/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/pow/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/remainder/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/remainder/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/round/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/round/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/rsqrt/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/rsqrt/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/sin/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/sin/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/sincos/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/sincos/Traits.hpp]: Typo: suppored instead of supported

File: include/alpaka/math/sqrt/Traits.hpp
Comment (j-stephan) [inline include/alpaka/math/sqrt/Traits.hpp]: Typo: suppored instead of supported

**Review by j-stephan — CHANGES_REQUESTED** [review]
Nice work! Requested changes are:

1. Clarification regarding @sbastrakov's question.
2. A typo that propagated through copy & paste.

File: include/alpaka/math/sqrt/Traits.hpp
Comment (bernhardmgruber) [reply include/alpaka/math/sqrt/Traits.hpp]: Are you bored? :D One comment would have been enough!

File: include/alpaka/math/abs/Traits.hpp
Comment (bernhardmgruber) [reply include/alpaka/math/abs/Traits.hpp:57]: The user calls `alpaka::math::abs(acc, value);` and the function on line 54 is picked. On line 57 we instantiate `traits::Abs<ImplementationBase, TArg>`. Notice that the template arguments are the backend and the type on which to compute the abs. If there is any specialization of `traits::Abs`, it will be picked. For `float` and `double`, such specializations exist for all backends. However, if there is no specialization, the primary template on line 33 will be instantiated which proceeds to do an ADL call.

We could document that somewhere. Where could be a good place?

**Comment by j-stephan** [comment]
Could you rebase on `develop`? These old CI tests are blocking my merge button.

**Comment by bernhardmgruber** [comment]
Rebased

**Comment by j-stephan** [comment]
Re-requesting review from @sbastrakov: Has your question been sufficiently answered?

File: include/alpaka/math/abs/Traits.hpp
Comment (sbastrakov) [reply include/alpaka/math/abs/Traits.hpp:57]: Thanks for explaining @bernhardmgruber . Where to document is a good question. I am not sure, maybe as a new subsection [here](https://alpaka.readthedocs.io/en/latest/dev/details.html). Btw I've also found out [the last point about naming here](https://alpaka.readthedocs.io/en/latest/dev/style.html#traits) is violated by this PR. However, I remember Benjamin himself recommending this new naming you used, so should be no issue there.

Comment (bernhardmgruber) [reply include/alpaka/math/abs/Traits.hpp:57]: As you said, @BenjaminW3 mentioned in some comment somewhere, that we should have picked `operator()` for trait implementations.
Furthermore, I had to change the function name to make it work. Consider this:
```c++
struct Acos
{
    ALPAKA_FN_HOST_ACC auto acos(T const& ctx, TArg const& arg) // instead of operator()
    {
        alpaka::ignore_unused(ctx);
        // This is an ADL call. If you get a compile error here then your type is not supported by the
        // backend and we could not find acos(TArg) in the namespace of your type.
        return acos(arg);
    }
};
```
<details>
  <summary>Guess what would happen :)</summary>
  An infinite recursion :D
</details>

**Review by sbastrakov — DISMISSED** [review]
My question was answered. I still think it's worth to add a brief explanation for documentation (see discussion above), but will leave it to @bernhardmgruber to decide.

**Comment by bernhardmgruber** [comment]
I have added documentation for this feature.

**Review by sbastrakov — APPROVED** [review]

**Review by psychocoderHPC — APPROVED** [review]

**Comment by psychocoderHPC** [comment]
@j-stephan Is your request resolved?

**Review by j-stephan — APPROVED** [review]

**Comment by j-stephan** [comment]
Yes. You're free to hit the merge button.

**Pull request merged**
